hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/inject/AvroUpsertTarget.java

```diff
@@ -40,60 +65,59 @@ class AvroUpsertTarget implements UpsertTarget {
     }
 
     @Override
+    @SuppressWarnings("ReturnCount")
     public UpsertInjector createInjector(@Nullable String path, QueryDataType type) {
         if (path == null) {
             return FAILING_TOP_LEVEL_INJECTOR;
         }
 
-        switch (type.getTypeFamily()) {
-            case TINYINT:
-                return value -> record.set(path, value == null ? null : ((Byte) value).intValue());
-            case SMALLINT:
-                return value -> record.set(path, value == null ? null : ((Short) value).intValue());
+        Schema fieldSchema = schema.getField(path).schema();
+        Schema.Type schemaFieldType = unwrapNullableType(fieldSchema).getType();
+        switch (schemaFieldType) {
             case BOOLEAN:
-            case INTEGER:
-            case BIGINT:
-            case REAL:
+            case INT:
+            case LONG:
+            case FLOAT:
             case DOUBLE:
-                return value -> record.set(path, value);
-            case DECIMAL:
-            case TIME:
-            case DATE:
-            case TIMESTAMP:
-            case TIMESTAMP_WITH_TIME_ZONE:
-            case VARCHAR:
-                return value -> record.set(path, QueryDataType.VARCHAR.convert(value));
-            case OBJECT:
-                return createObjectInjector(path);
+            case STRING:
+                return value -> {
+                    try {
+                        record.set(path, AVRO_TO_SQL.get(schemaFieldType).convert(value));
+                    } catch (QueryException e) {
+                        throw QueryException.error("Cannot convert " + value + " to " + schemaFieldType
+                                + " (field=" + path + ")");
+                    }
+                };
+            case UNION:
+                return value -> {
+                    if (value == null) {
+                        record.set(path, null);
+                        return;
+                    } else {
+                        for (Schema.Type target : CONVERSION_PREFS.get(value.getClass())) {
+                            if (fieldSchema.getTypes().stream().anyMatch(schema -> schema.getType() == target)) {
+                                try {
+                                    record.set(path, AVRO_TO_SQL.get(target).convert(value));
```

Comment:
`record.set` does not check type of the passed value for the field. Can we produce garbage this way?

Reply:
`GenericData` has methods `validate(Schema, Object)` and `resolveUnion(Schema, Object)`. The latter is used when writing to a union, but the former doesn't seem to be used anywhere. We can use that method to verify the conversion result, but converters are expected to produce objects in the correct type.